test-suite/src/test/scala/org/scalajs/testsuite/library/UnionTypeTest.scala

```diff
@@ -0,0 +1,153 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package org.scalajs.testsuite.library
+
+import scala.language.implicitConversions
+
+import scala.scalajs.js
+import js.|
+
+import org.scalajs.jasminetest.JasmineTest
+import org.scalajs.testsuite.Typechecking._
+
+import js.JSConverters._
+
+object UnionTypeTest extends JasmineTest {
+
+  private implicit def unionAsJSAny(x: _ | _): js.Any = x.asInstanceOf[js.Any]
```

Comment:
This is not implementable, unfortunately. Because when you have a `A | B`, you don't know whether `x` is an `A` or a `B`, and you have no means of testing which one it is, in general. So you cannot know whether you should apply the conversion from `A` to `js.Any` or the conversion from `B` to `js.Any`.

This is also why the implicit evidences are all "fake", instead of having an `apply` method that "convert" the values (even to just cast in most cases). Because the `allSubtypes` evidence wouldn't know how to construct its `apply` method. This, in turn, is why `Evidence` is sealed. Because we cannot allow "custom" conversions that are not known to boil down to a hard cast.